# Typecast

* Implicit Type casting
* Explicit Type casting

### Implicit Type Conversion

In [1]:
a = 10       # a is an integer
b = 10.1     # b is a float

print(a)
print(b)

10
10.1


In [2]:
print(type(a))   
print(type(b))

<class 'int'>
<class 'float'>


In [3]:
# Lets add the two numbers variable a and variable b
# and save the output as a new variable c

c = a+b
print(c)

20.1


In [4]:
print(type(c))

<class 'float'>


As you have seen, the addition of 10 and 10.1 is a decimal (float) 20.1

Of course, it should be 20.1, a float value, right. Actually, its slightly more complicated than one might think. 10 is an integer and 10.1 is a float data type.

To be honest, we have different data types for a reason. By default we can not add one type of data type to another one. But because addding up integers and floats outputs a decimal, python first does a type conversion of integer to float.

So, python first converts the data type of variable a (integer 10) to float  and then the two float values (a+b) are added, so the result comes as a data type of float. This process when the type conversion is done automatically by python is called **implicit type casting**.


Some little details that we should we aware of, right. Not convinced yet? Let's look at more challenging example. 



In [1]:
# This is supposed to throw an error as we will attempt 
# to add a string to a float

# a = "10" 

# b = 10.1

# The below statement will throw an error
# c = a+b

# print(c)

The value of string "10" looks like its a number, so intuitively we might think that we can add this string "10" to a float 10.1 but like you noticed, these two variables had different data type. A string data type can not be added to a float (or integer) data type.

But because we know that the value of the string is "10", we can convert this string of value "10" to an integer first and this sort of type conversion is called **explicit type conversion**

In [6]:
print(a)

10


In [7]:
print(type(a))

<class 'str'>


Here we'll use the int() inbuilt python function to explicitly convert the data type of variable a from string to integer.

In [8]:
d = int(a)
print(d)
print(type(d))

10
<class 'int'>


Now, we are ready to add the integer variable d and float variable b. 

In [10]:
print(d + b)

20.1


Again, one implicit type conversion happened, which first converted data type of variable d to float before adding variable d to variable b.

**Note**: Type casting should be used with caution as not all strings could be converted into integers or floats.

In [13]:
e = "2 apples"
print(e)
print(type(e))

2 apples
<class 'str'>


Because our variable e contains characters now, the whole string can not be converted to integer.

In [1]:
# The following code will give an error
# int(e)

## Type conversion of booleans

In [15]:
f = True
g = False

Remember, True boolean values are stored as 1 in the backend of python or any programming language.

In [18]:
print(int(f))

1


False boolean values are stored as 0 in the backend of python.

In [19]:
print(int(g))

0


Because booleans and numerics are so related to each other, we can use type conversion between booleans and numeric values.

In [20]:
bool(1)

True

In [21]:
bool(0)

False

Actually, if we convert any numeric value other than 0 to boolean, python will be able to convert (Type cast) it into True boolean.

If the number is zero, then only python converts it into False.

In [26]:
bool(1.1)

True

In [27]:
bool(0.0)

False

In [28]:
bool(0.1)

True

Even the negative numbers are True boolean

In [30]:
bool(-10)

True

In [31]:
bool(-10.33)

True

-0 and 0 means the same thing, so these are two of the only 3 cases where False comes as output

In [33]:
bool(-0)

False

**None** data type means a null value, a value that does not exist. So, if there's nothing, upon type conversion python converts that into False .

In [35]:
bool(None)

False